### PR TITLE
zcash_client_sqlite: Read rcm correctly from data DB after Canopy

### PR DESCRIPTION
**Pull request by str4d**
ZIP 212 alters the note plaintext to store a seed from which rcm is
derived, rather than storing rcm directly. In the mobile SDKs we only
need rcm, so for post-ZIP 212 notes, we derive rcm from the seed and
store rcm in the data DB.

However, when selecting notes to spend, `create_to_address` was using the
transaction's target height to determine if Canopy is active, and parsing
the rcm value as the seed if so. This effectively applied a seed->rcm
derivation to all selected notes' rcms once Canopy activated on the
chain. As a result, the note commitments were incorrect, and thus the
anchors derived from the witness paths were also incorrect. This caused
two kinds of observed failures:

- If more than one note was selected, the builder would fail with
  "anchor mismatch", as the note commitments would be effectively
  randomised, causing the derived anchors to also randomise.
- If a single note was selected, the transaction would be built using
  the randomised anchor, and then rejected when sent to the network.

The fix is to "pretend" in `create_to_address` that all notes are
pre-ZIP 212 notes. This works fine because we never need to serialize
back to the note plaintext while spending a note.

### CODE REVIEW

**Comment by str4d** [comment]
This will fix https://github.com/zcash/ZcashLightClientKit/issues/212 once `zcash_client_sqlite 0.2.1` is published.

**Comment by str4d** [comment]
I based this PR off 37b5d67d33216f3717f8ad7ea1a2fbc8a3a0b548 (the `zcash_client_sqlite 0.2.0` commit). I'll be publishing 74bf85641cdc0f596c6ec378ed92bc4c157447bd once this PR is reviewed; the merge commit resolves subsequent changes from master so that this PR can be merged.

**Comment by adityapk00** [comment]
Nice catch. I think Zecwallet is probably ok, because I’m also serializing the type of note. Can someone cast an eye on this:

https://github.com/adityapk00/zecwallet-light-cli/blob/master/lib/src/lightwallet/data.rs#L74

**Comment by adityapk00** [comment]
Also, does this mean that pre canopy and post canopy notes cannot be mixed and spent in the same transaction? Or did I misunderstand?

**Comment by str4d** [comment]
They can be freely mixed. All that matters is you know how to obtain `rcm` for each note (either directly, or from `rseed`), so that the correct note commitment is calculated during proving.

**Review by therealyingtong — APPROVED** [review]
untested ACK

File: zcash_client_sqlite/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to Rust's notion of
+[Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [0.2.1] - 2020-10-24
+### Fixed
+- `transact::create_to_address` now correctly reconstructs notes from the data
+  DB after Canopy activation (zcash/librustzcash#311).
```
Comment (daira) [inline zcash_client_sqlite/CHANGELOG.md:11]:
```suggestion
  DB after Canopy activation (zcash/librustzcash#311). This is critcal to correct
  operation of spends after Canopy.
```

File: zcash_client_sqlite/src/transact.rs [hunk 2]
```diff
@@ -230,19 +230,16 @@ pub fn create_to_address<DB: AsRef<Path>, P: consensus::Parameters>(
             let rseed = {
                 let d: Vec<_> = row.get(2)?;
```
Comment (daira) [inline zcash_client_sqlite/src/transact.rs:231]:
```suggestion
                let rcm_bytes: Vec<_> = row.get(2)?;
```

File: zcash_client_sqlite/src/transact.rs [hunk 3]
```diff
@@ -230,19 +230,16 @@ pub fn create_to_address<DB: AsRef<Path>, P: consensus::Parameters>(
...
+                // We store rcm directly in the data DB, regardless of whether the note
+                // used a v1 or v2 note plaintext, so for the purposes of spending let's
+                // pretend this is a pre-ZIP 212 note.
+                let r = jubjub::Fr::from_repr(
+                    d[..]
+                        .try_into()
+                        .map_err(|_| Error(ErrorKind::InvalidNote))?,
+                )
+                .ok_or(Error(ErrorKind::InvalidNote))?;
+                Rseed::BeforeZip212(r)
```
Comment (daira) [inline zcash_client_sqlite/src/transact.rs:242]:
```suggestion
                let rcm = jubjub::Fr::from_repr(
                    rcm_bytes[..]
                        .try_into()
                        .map_err(|_| Error(ErrorKind::InvalidNote))?,
                )
                .ok_or(Error(ErrorKind::InvalidNote))?;
                Rseed::BeforeZip212(rcm)
```

**Review by daira — APPROVED** [review]
utACK with non-blocking suggestions.

**Comment by daira** [comment]
@adityapk00 wrote:
> Nice catch. I think Zecwallet is probably ok, because I’m also serializing the type of note. Can someone cast an eye on this:
>
> https://github.com/adityapk00/zecwallet-light-cli/blob/master/lib/src/lightwallet/data.rs#L74

That code looks correct to me. Which code are you using to decrypt notes in the first place?

File: zcash_client_sqlite/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to Rust's notion of
+[Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [0.2.1] - 2020-10-24
+### Fixed
+- `transact::create_to_address` now correctly reconstructs notes from the data
+  DB after Canopy activation (zcash/librustzcash#311). This is critcal to correct
```
Comment (daira) [inline zcash_client_sqlite/CHANGELOG.md:11]: Oops, "critcal". Oh well.

**Comment by adityapk00** [comment]
> Which code are you using to decrypt notes in the first place?

I'm just using `librustzcash`'s methods to do note decryption. 
The compact note decryption: https://github.com/adityapk00/zecwallet-light-cli/blob/master/lib/src/lightwallet.rs#L1518
Full note decryption when fetching full txns: https://github.com/adityapk00/zecwallet-light-cli/blob/master/lib/src/lightwallet.rs#L1327

**Pull request merged**